### PR TITLE
Upgrade to Jandex 3.0.0

### PR DESCRIPTION
**Pull request by Ladicek**
Submitting as draft because this needs discussion on release coordination (I guess).

Currently depends on SNAPSHOT of SmallRye GraphQL.

### CODE REVIEW

**Comment by gastaldi** [comment]
@Ladicek don't forget https://github.com/quarkusio/quarkus/pull/27520

**Comment by gastaldi** [comment]
Do you think it's interesting to add the org.jboss.jandex as a banned dependency in the enforcer plugin (build-parent)?

**Comment by Ladicek** [comment]
Thanks, fixed `independent-projects/bootstrap`. The Enforcer plugin is already configured to ban `org.jboss:jandex` in this PR -- it is a recommendation I make in the Jandex 3.0.0 [release announcement](https://smallrye.io/blog/jandex-3-0-0/), so I'm well aware of it :-)

**Comment by Sanne** [comment]
Depends on https://github.com/quarkusio/quarkus/pull/27602

File: bom/application/pom.xml [hunk 1]
```diff
@@ -4830,6 +4840,12 @@
                 <groupId>org.hibernate</groupId>
                 <artifactId>hibernate-core</artifactId>
                 <version>${hibernate-orm.version}</version>
+                <exclusions>
+                    <exclusion>
+                        <groupId>org.jboss</groupId>
+                        <artifactId>jandex</artifactId>
+                    </exclusion>
```
Comment (Sanne) [inline bom/application/pom.xml:4847]: :+1:

**Review by gsmet — DISMISSED** [review]
Reminder: we will need corresponding Jakarta updates in `jakarta/rewrite.yml` for SmallRye OpenAPI and SmallRye GraphQL.

**Comment by Ladicek** [comment]
After incorporating #27769, the only remaining SNAPSHOT is SmallRye GraphQL. I've discussed with @jmartisk on Zulip and the plan is to release SmallRye GraphQL roughly in a week or so.

**Comment by jmartisk** [comment]
If you prefer we can do it earlier, I just wanted to avoid the risk of doing a new minor SmallRye release and then finding out we want to add some major new stuff ASAP (into Quarkus 2.13 through SmallRye). But I'm not aware of any such incoming stuff (maybe @phillip-kruger knows something more). I appreciate that a major Jandex upgrade warrants some bake time before Quarkus gets released with it.

**Comment by gsmet** [comment]
The Jandex upgrade will be merged after I branch 2.13 CR1 so on September 14th. So the idea is to have the pull request ready for that date (or just after).

**Comment by jmartisk** [comment]
Ok then we'll aim for releasing SmallRye on, say, Monday 12th

**Comment by Ladicek** [comment]
No I think the original agreement still holds. This PR targets Quarkus 2.14, so SmallRye GraphQL release around the point of branching off 2.13 is still fine. Thank you!

**Comment by jmartisk** [comment]
Yea I forgot this was targeting 2.14, sorry

**Comment by gsmet** [comment]
@jmartisk @Ladicek  `main` is now `2.14`, so we can consider merging this when it's ready. IIRC, we need SmallRye GraphQL releases.

**Comment by Ladicek** [comment]
Correct, still waiting for SmallRye GraphQL 1.8.0.

**Comment by jmartisk** [comment]
Ok, I'll go for the release now

**Comment by gsmet** [comment]
@jmartisk I think you have that in mind but since you used the singular: we also need a Jakarta release.

**Comment by jmartisk** [comment]
SmallRye GraphQL 2.0.0.RC9 and 1.8.0 are in Central.

File: jakarta/rewrite.yml [hunk 2]
```diff
@@ -585,7 +585,7 @@ recipeList:
       newValue: 4.0.0-RC1
```
Comment (jmartisk) [inline jakarta/rewrite.yml:585]: Please update to `4.0.0`

Comment (Ladicek) [reply jakarta/rewrite.yml:585]: That should be a different PR I guess?

Comment (Ladicek) [reply jakarta/rewrite.yml:585]: Ah if GraphQL depends on it, then we can do that here as well, I think.

Comment (jmartisk) [reply jakarta/rewrite.yml:585]: I don't know, depends how hard we try to keep dependencies properly aligned :)

**Comment by Ladicek** [comment]
OK, marking as ready for review. The 1st commit deserves a review from @mkouba (though this change won't be necessary when I release Jandex 3.0.1, which should restore structural `Type` equality).

**Comment by Ladicek** [comment]
Note that the "quickstarts compilation" check in CI will fail due to:

- Reactive Messaging HTTP incompatibility with Jandex 3 (fixed in 1.0.7, https://github.com/quarkiverse/quarkus-reactive-messaging-http/pull/92)
- OptaPlanner Quarkus extension incompatibility with Jandex 3 (https://github.com/kiegroup/optaplanner/pull/2193)

That's all I'm currently aware of.

**Comment by gsmet** [comment]
@mkouba can you have a look at the first commit? I'd like to get this merged ASAP. Thanks!

**Comment by mkouba** [comment]
> OK, marking as ready for review. The 1st commit deserves a review from @mkouba (though this change won't be necessary when I release Jandex 3.0.1, which should restore structural `Type` equality).

The first commit looks good. If there are problems we can revert the change once Jandex 3.0.1 is released.

**Review by mkouba — APPROVED** [review]

**Review by gsmet — APPROVED** [review]
Let's get this in! Thanks for your work on this @Ladicek !

**Pull request merged**